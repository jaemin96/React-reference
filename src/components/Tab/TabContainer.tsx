import React, { ReactNode, useState } from 'react';
import { TabContent, TabItem, TabMenu } from './';

interface TabContainerProps {
  activeColor?: string;
  activeBgColor?: string;
  inactiveColor?: string;
  inactiveBgColor?: string;
  fontSize?: string;
  borderColor?: string;
  children: ReactNode;
}

export const TabContainer = (props: TabContainerProps) => {
  const { children, activeColor, activeBgColor, inactiveColor, inactiveBgColor, fontSize, borderColor } = props;
  const [activeTab, setActiveTab] = useState<number>(0);

  const handleTab = (index: number) => {
    setActiveTab(index);
  };

  return (
    <div className="tab-container" data-testid="tab">
      <TabMenu borderColor={borderColor}>
        {React.Children.map(children, (child, index) => {
          if (React.isValidElement(child)) {
            return (
              <TabItem
                activeColor={activeColor}
                activeBgColor={activeBgColor}
                inactiveColor={inactiveColor}
                inactiveBgColor={inactiveBgColor}
                borderColor={borderColor}
                fontSize={fontSize}
                key={index}
                index={index}
                isActive={index === activeTab}
                onClick={() => handleTab(index)}
              >
                {child.props.label}
              </TabItem>
            );
          }
          return null;
        })}
      </TabMenu>
      <TabContent>
        {React.Children.map(children, (child, index) =>
          React.isValidElement(child) && index === activeTab ? child : null
        )}
      </TabContent>
    </div>
  );
};
