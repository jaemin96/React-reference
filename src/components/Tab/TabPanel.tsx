import { ReactNode } from 'react';

interface TabPanelProps {
  label: string;
  children: ReactNode;
}

export const TabPanel = ({ children }: TabPanelProps) => {
  return <div className="tab-panel">{children}</div>;
};
