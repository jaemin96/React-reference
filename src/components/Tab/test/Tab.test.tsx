import { render, screen, fireEvent } from '@testing-library/react';
import { TabContainer } from '../TabContainer';
import { TabPanel } from '../TabPanel';

describe('Tab', () => {
  beforeEach(() => {
    render(
      <TabContainer>
        <TabPanel label="tab1">Tab 1 content</TabPanel>
        <TabPanel label="tab2">Tab 2 content</TabPanel>
        <TabPanel label="tab3">Tab 3 content</TabPanel>
      </TabContainer>
    );
  });

  test('render tab component', () => {
    const tabElement = screen.getByTestId('tab');
    expect(tabElement).toBeInTheDocument();
  });

  test('check init status', () => {
    const tabItemElement = screen.getByTestId('tabItem-1');
    expect(tabItemElement).toHaveClass('active');
  });

  test('check tab change', () => {
    const tabItem1 = screen.getByTestId('tabItem-1');
    const tabItem2 = screen.getByTestId('tabItem-2');

    fireEvent.click(tabItem2);
    expect(tabItem2).toHaveClass('active');

    fireEvent.click(tabItem1);
    expect(tabItem1).toHaveClass('active');
  });

  test('snapshot', () => {
    const { asFragment } = render(
      <TabContainer>
        <TabPanel label="tab1">Tab 1 content</TabPanel>
        <TabPanel label="tab2">Tab 2 content</TabPanel>
        <TabPanel label="tab3">Tab 3 content</TabPanel>
      </TabContainer>
    );
    expect(asFragment()).toMatchSnapshot();
  });
});

// describe("tab2" , () => {
//     test("not exist children", () => {
//         render(
//             <TabContainer>
//             </TabContainer>
//         );
//         const tabMenuElement = screen.getByRole('tabMenu');
//         expect(tabMenuElement).toBeNull();
//     })
// })
