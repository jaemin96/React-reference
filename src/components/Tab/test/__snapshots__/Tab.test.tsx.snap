// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tab snapshot 1`] = `
<DocumentFragment>
  <div
    class="tab-container"
    data-testid="tab"
  >
    <div
      class="tab-menu"
      data-testid="tabMenu"
    >
      <button
        class="tab-item active"
        data-testid="tabItem-1"
      >
        tab1
      </button>
      <button
        class="tab-item "
        data-testid="tabItem-2"
      >
        tab2
      </button>
      <button
        class="tab-item "
        data-testid="tabItem-3"
      >
        tab3
      </button>
    </div>
    <div
      class="tab-content"
      data-testid="tabContent"
    >
      <div
        class="tab-panel"
      >
        Tab 1 content
      </div>
    </div>
  </div>
</DocumentFragment>
`;
